src/data_source/fetching.rs

```diff
@@ -641,22 +738,84 @@ where
             .boxed()
     }
 
-    async fn store<S>(self, storage: &mut NotifyStorage<Types, S>) -> Result<(), S::Error>
-    where
-        S: UpdateAvailabilityData<Types> + Send + Sync,
+    async fn active_fetch<S, P>(
+        fetcher: Arc<Fetcher<Types, S, P>>,
+        storage: &RwLockReadGuard<'_, NotifyStorage<Types, S>>,
+        req: Self::Request,
+    ) where
+        S: AvailabilityStorage<Types> + 'static,
+        P: AvailabilityProvider<Types>,
     {
-        storage.leaf_notifier.notify(&self);
-        storage.storage.insert_leaf(self).await
+        fetch_leaf_with_callbacks(fetcher, storage, req, None)
     }
 
-    async fn load<S>(storage: &S, req: Self::Request) -> QueryResult<Self>
+    async fn load<S>(storage: &NotifyStorage<Types, S>, req: Self::Request) -> QueryResult<Self>
     where
         S: AvailabilityStorage<Types>,
     {
-        storage.get_leaf(req).await
+        // Fail quickly, without touching storage, if the requested height is greater than the
+        // current height. In this case, we know we don't have the leaf.
+        if let LeafId::Number(n) = req {
+            ensure!((n as u64) < storage.height, NotFoundSnafu);
+        }
+        storage.storage.get_leaf(req).await
+    }
+}
+
+fn fetch_leaf_with_callbacks<Types, S, P, I>(
+    fetcher: Arc<Fetcher<Types, S, P>>,
+    storage: &RwLockReadGuard<'_, NotifyStorage<Types, S>>,
+    req: LeafId<Types>,
+    callbacks: I,
+) where
+    Types: NodeType,
+    Payload<Types>: QueryablePayload,
+    S: AvailabilityStorage<Types> + 'static,
+    P: AvailabilityProvider<Types>,
+    I: IntoIterator<Item = LeafCallback<Types, S, P>> + Send + 'static,
+    I::IntoIter: Send,
+{
+    match req {
+        LeafId::Number(n) => {
+            let height = storage.height as usize;
+            if n >= height {
+                // If the requested leaf has yet to be produced, based on the current block height,
+                // there is no point in requesting it. We will receive it passively once it is
+                // created.
+                tracing::info!("not fetching leaf {n} because height is only {height}");
+                return;
+            }
+
+            let fetcher = fetcher.clone();
+            fetcher.leaf_fetcher.clone().spawn_fetch(
+                n.into(),
+                fetcher.provider.clone(),
+                once(LeafCallback::Leaf { fetcher }).chain(callbacks),
```

Comment:
We use a dedicated type/trait for callbacks now, rather than a closure, because callbacks actually have an identity. More about this in `src/fetching.rs`, but basically this is to ensure that all callbacks run, for example in the case where we're fetching a leaf and the corresponding block at the same time, even though we only actually do the fetch once.